data/transactions/logic/debugger_eval_test.go

```diff
@@ -0,0 +1,303 @@
+	"github.com/algorand/go-algorand/data/transactions/logic/mocktracer"
+	"github.com/algorand/go-algorand/test/partitiontest"
+	"github.com/stretchr/testify/require"
+)
+
+const debuggerTestProgramApprove string = `intcblock 0 1 1 1 1 5 100
+bytecblock 0x414c474f 0x1337 0x2001 0xdeadbeef 0x70077007
+bytec 0
+sha256
+keccak256
+sha512_256
+len
+intc_0
++
+intc_1
+-
+intc_2
+/
+intc_3
+*
+intc 4
+<
+intc_1
+>
+intc_1
+<=
+intc_1
+>=
+intc_1
+&&
+intc_1
+||
+bytec_1
+bytec_2
+!=
+bytec_3
+bytec 4
+!=
+&&
+&&
+`
+const debuggerTestProgramReject string = debuggerTestProgramApprove + "!"
+const debuggerTestProgramError string = debuggerTestProgramApprove + "err"
+
+func TestWebDebuggerManual(t *testing.T) { //nolint:paralleltest // Manual test
+	partitiontest.PartitionTest(t)
+
+	debugURL := os.Getenv("TEAL_DEBUGGER_URL")
+	if len(debugURL) == 0 {
+		t.Skip("this must be run manually")
+	}
+
+	ep, tx, _ := MakeSampleEnv()
+	ep.TxnGroup[0].Lsig.Args = [][]byte{
+		tx.Sender[:],
+		tx.Receiver[:],
+		tx.CloseRemainderTo[:],
+		tx.VotePK[:],
+		tx.SelectionPK[:],
+		tx.Note,
+	}
+	ep.Tracer = MakeEvalTracerDebuggerAdaptor(&WebDebugger{URL: debugURL})
+	TestLogic(t, debuggerTestProgramApprove, AssemblerMaxVersion, ep)
+}
+
+type testDebugger struct {
+	register int
+	update   int
+	complete int
+	state    *DebugState
+}
+
+func (d *testDebugger) Register(state *DebugState) {
+	d.register++
+	d.state = state
+}
+
+func (d *testDebugger) Update(state *DebugState) {
+	d.update++
+	d.state = state
+}
+
+func (d *testDebugger) Complete(state *DebugState) {
+	d.complete++
+	d.state = state
+}
+
+func TestDebuggerProgramEval(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	simpleTestCases := []struct {
+		name             string
+		program          string
+		evalProblems     []string
+		expectedRegister int
+		expectedUpdate   int
+		expectedComplete int
+		expectedStack    []basics.TealValue
+	}{
+		{
+			name:             "approve",
+			program:          debuggerTestProgramApprove,
+			expectedRegister: 1,
+			expectedUpdate:   35,
+			expectedComplete: 1,
+			expectedStack: []basics.TealValue{
+				{
+					Type: basics.TealUintType,
+					Uint: 1,
+				},
+			},
+		},
+		{
+			name:             "reject",
+			program:          debuggerTestProgramReject,
+			evalProblems:     []string{"REJECT"},
+			expectedRegister: 1,
+			expectedUpdate:   36,
+			expectedComplete: 1,
+			expectedStack: []basics.TealValue{
+				{
+					Type: basics.TealUintType,
+					Uint: 0,
+				},
+			},
+		},
+		{
+			name:             "error",
+			program:          debuggerTestProgramError,
+			evalProblems:     []string{"err opcode executed"},
+			expectedRegister: 1,
+			expectedUpdate:   36,
+			expectedComplete: 1,
+			expectedStack: []basics.TealValue{
+				{
+					Type: basics.TealUintType,
+					Uint: 1,
+				},
+			},
+		},
+	}
+
+	t.Run("logicsig", func(t *testing.T) {
+		t.Parallel()
+		for _, testCase := range simpleTestCases {
```

Comment:
I do this sometimes too, so I'm of two minds, but:

Would it be better to make the test cases global variables, and have these `t.Run` calls as separate top-level tests? My thinking is that if they are independent, it would be nice to have them run independently, and also easier to run a single one by hand when debugging a problem.

Of course, if there are dependencies between them, it should certainly run this way.

Reply:
Good point, resolved in ac2883d